torchtext/prototype/t5/modules.py

```diff
@@ -0,0 +1,819 @@
+    Args:
+        encoder_layer: an instance of the TransformerEncoderLayer() class (required).
+        num_layers: the number of sub-encoder-layers in the encoder (required).
+        norm: the layer normalization component (optional).
+        enable_nested_tensor: if True, input will automatically convert to nested tensor
+            (and convert back on output). This will improve the overall performance of
+            TransformerEncoder when padding rate is high. Default: ``True`` (enabled).
+    Examples::
+        >>> encoder_layer = T5EncoderLayer(d_model=768, nhead=12, dim_feedfoward=3072, dropout=0.1, activation='relu', batch_first=True)
+        >>> t5_norm = T5LayerNorm(d_model=768)
+        >>> t5_encoder = T5Encoder(encoder_layer, num_layers=12, norm=t5_norm)
+        >>> src = torch.rand(10, 32, 512)
+        >>> out = t5_encoder(src)
+    """
+
+    def __init__(
+        self,
+        encoder_layer,
+        relative_attention_num_buckets,
+        num_heads,
+        num_layers=12,
+        norm=None,
+        enable_nested_tensor=True,
+    ):
+        super(T5Encoder, self).__init__(encoder_layer, num_layers, norm, enable_nested_tensor)
+
+        first_layer = copy.deepcopy(encoder_layer)
+        first_layer.compute_relative_attention_bias = True
+        first_layer.relative_attention_bias = nn.Embedding(relative_attention_num_buckets, num_heads)
+        self.layers = nn.ModuleList([first_layer] + [copy.deepcopy(encoder_layer) for i in range(num_layers - 1)])
+        self.num_layers = num_layers
+        self.enable_nested_tensor = enable_nested_tensor
+
+    def forward(
+        self, src: Tensor, mask: Optional[Tensor] = None, src_key_padding_mask: Optional[Tensor] = None
+    ) -> Tensor:
+        r"""Pass the input through the encoder layers in turn.
+        Args:
+            src: the sequence to the encoder (required).
+            mask: the mask for the src sequence (optional).
+            src_key_padding_mask: the mask for the src keys per batch (optional).
+        Shape:
+            see the docs in Transformer class.
+        """
+
+        output = src
+        position_bias = None
+        for mod in self.layers:
+            output, position_bias = mod(
+                output, src_mask=mask, src_key_padding_mask=src_key_padding_mask, position_bias=position_bias
+            )
+
+        return output
+
+
+class T5EncoderModel(nn.Module):
+    def __init__(
+        self,
+        d_model: int,
+        d_feedforward: int,
+        dropout: float,
+        activation: Union[str, Callable[[Tensor], Tensor]],
+        layer_norm_eps: float,
+        num_heads: int,
+        num_layers: int,
+        batch_first: bool,
+        relative_attention_num_buckets: int,
+        relative_attention_max_distance: int,
+        padding_idx: int,
+        max_seq_len: int,
+        vocab_size: int,
+    ) -> None:
+        super().__init__()
```

Comment:
nit: Let's add docstrings explaining the purpose of the class and all of the input arguments before we actually merge in the implementation